features/FEATURE_BLE/ble/BLETypes.h

```diff
@@ -117,6 +119,266 @@ static inline attribute_handle_range_t attribute_handle_range(
+     */
+    PasskeyAscii(passkey_num_t passkey) {
+        for (int i = 5, m = 100000; i >= 0; --i, m /= 10) {
+            uint32_t result = passkey / m;
+            ascii[i] = NUMBER_OFFSET + result;
+            passkey -= result;
+        }
+    }
+
+    /**
+     * Cast to number.
+     */
+    operator passkey_num_t() {
+        return to_num(ascii);
+    }
+
+    /**
+     * Convert ASCII string of digits into a number.
+     * @param ASCII string of 6 digits stored as ASCII characters
+     * @return Passkey as a number.
+     */
+    static uint32_t to_num(const uint8_t *ascii) {
+        uint32_t passkey = 0;
+        for (size_t i = 0, m = 1; i < PASSKEY_LEN; ++i, m *= 10) {
+            passkey += (ascii[i] - NUMBER_OFFSET) * m;
+        }
+        return passkey;
+    }
+
+    /**
+     * Return the pointer to the buffer holding the string.
+     */
+    uint8_t* value() {
+        return ascii;
+    }
+private:
+    uint8_t ascii[PASSKEY_LEN];
+};
+
+template <size_t octet_size>
+struct octet_type_t {
+    /**
+     * Default to all zeroes
+     */
+    octet_type_t() {
+        memset(_value, 0x00, sizeof(_value));
+    }
+
+    /**
+     * Initialize a data from an array of bytes.
+     *
+     * @param input_value value of the data.
+     */
+    octet_type_t(const uint8_t *input_value) {
+        memcpy(_value, input_value, sizeof(_value));
+    }
+
+    /**
+     * Initialize a data from an buffer of bytes.
+     *
+     * @param input_value pointer to buffer.
+     * @param size buffer size
+     */
+    octet_type_t(const uint8_t* input_value, size_t size) {
+        memcpy(_value, input_value, size);
+    }
+
+    /**
+     * Equal operator between two octet types.
+     */
+    friend bool operator==(const octet_type_t& lhs, const octet_type_t& rhs) {
+        return memcmp(lhs._value, rhs._value, sizeof(lhs._value)) == 0;
+    }
+
+    /**
+     * Non equal operator between two octet types.
+     */
+    friend bool operator!=(const octet_type_t& lhs, const octet_type_t& rhs) {
+        return !(lhs == rhs);
+    }
+
+    /**
+     * Subscript operator to access data content
+     */
+    uint8_t& operator[](uint8_t i) {
+        return _value[i];
+    }
+
+    /**
+     * Return the pointer to the buffer holding data.
+     */
+    const uint8_t* data() const {
+        return _value;
+    }
+
+    /**
+     * Return the pointer to the buffer holding data.
+     */
+    uint8_t* buffer() {
+        return _value;
+    }
+
+    /**
+     * Size in byte of a data.
+     */
+    static size_t size() {
+        return octet_size;
+    }
+
+protected:
+    uint8_t _value[octet_size];
+};
+
+/** 128 bit keys used by paired devices */
+typedef octet_type_t<16> key_t;
+typedef octet_type_t<16> irk_t;
+typedef octet_type_t<16> csrk_t;
+typedef octet_type_t<16> ltk_t;
+
+/** Used to identify LTK for legacy pairing connections */
+typedef octet_type_t<2> ediv_t;
+typedef octet_type_t<8> rand_t;
+
```

Comment:
`rand_t`, `random_data_t` and `oob_rand_t` have confusing names/overlapping uses and don't necessarily need to be separate.
I suggest:
`rand_t`, `random_data_t` --> `rand8_t`
`oob_rand_t` --> `rand16_t`

If you want to keep this split you can add:
```C
typedef rand8_t legacy_pairing_rand_t;
typedef rand16_t  oob_rand_t;
```

Reply:
`ediv` and `rand` have specific meaning in the security manger key distribution context. Having `rand8_t` and `rand16_t` can be confusing. 

I'm not sure that `random_data_t` actually needs a typedef; it is filled by the function `get_random_data` and it should be ok to pass in an array of 8 bytes.

I'd suggest to rename `oob_rand_t` into `oob_lesc_value_t` to be more inline with the documentation.

Reply:
The types are documented and have different applications. I don't have anything against lesc_value_t, I named it random for internal consistency, it's called different things in the spec at different times. Will rename.

Random data is just for convenience and type checking.

Reply:
OK, let's have `octet_type_t<8>` as the output type of `get_random_data()`. For `rand_t` and `ediv_t` I understand the matching with the spec but would be helpful having a hint with a prefix (e.g. `legacy_pairing_rand_t`,` legacy_pairing_ediv_t` or even shorter `lp_rand_t`,` lp_ediv_t`). `oob_lesc_value_t ` sounds good.

Reply:
we don't use the prefix for other types, and the type is unique to legacy pairing, there is no secure connections ediv and rand

of course this is now no longer true after the renaming of oob random to oob_lesc_value but I wouldn't want to go adding prefixes to everything

Reply:
removed random_data in favour of octet_type_t<8>